### ISSUE TITLE
Further distances on pitched heatmaps "flicker" on every frame

### ISSUE THREAD

**Issue by lbud** [mapbox/mapbox-gl-js#5563]
Pitching maps with heatmap layers cause "far" kernels to change on every frame (kind of flicker, in a blurry heatmap sort of way).

![heatmap-blur](https://user-images.githubusercontent.com/3170312/32201865-55979158-bd96-11e7-9cc9-6a03a2fdc6bc.gif)

Reproducible on any heatmap, including the [examples page](https://www.mapbox.com/mapbox-gl-js/example/heatmap-layer/).
cc @mourner

**Comment by mourner** [comment]
Wow, not sure what's going on there....

**Comment by lbud** [comment]
@mourner the effect sort of reminds me of something I ran into in fill-extrusion layers in gl-native, where sometimes the depth buffer resolution was too low (https://github.com/mapbox/mapbox-gl-native/pull/8431#issuecomment-296860060), causing jagged edges on the sides of extrusions; we fixed this by setting the near clip plan in the fill-extrusion layer projection matrix to 100 so as not to waste a lot of precision on shapes extremely close to the camera (https://github.com/mapbox/mapbox-gl-native/pull/8431/commits/72278bfe94ce0d19b515bb92e9d0be88eb53aa89). Not sure if this is relevant to what's going on here, but 🤷‍♀️

**Comment by wysisoft** [comment]
I don't think this is as much a "further distances" problem as much as a "small radius" problem. Radii above 5 seems to behave well. Radii 3 and under seem to flicker.

**Comment by pathmapper** [comment]
Agree with @wysisoft, there is no need to pitch a heatmap to trigger the bug. This GIF is just the official example zoomed out so that `heatmap-radius = 2`.

![heatmap-flicker](https://user-images.githubusercontent.com/20856381/68531859-34038b00-0317-11ea-8ba0-43c81158a4b5.gif)

**Comment by ryanhamley** [comment]
Thanks for pointing out what's going on here @pathmapper and @wysisoft

I can confirm that this is reproducible in our debug heatmap page as well, if you lower the `heatmap-radius` to <= 2. I'm not very familiar with the heatmap code though. @mourner does this new information point to a potential cause? could it be related to texture filtering a la https://github.com/mapbox/mapbox-gl-js/pull/8738?

**Comment by mpulkki-mapbox** [comment]
This bug is an aliasing issue caused by too low texture resolution in heatmap rendering. Texture size is 1/16 of the viewport size causing the "flickering". Here's an another example of the bug:
![heatmap_aliasing_bug](https://user-images.githubusercontent.com/55925868/68945343-de1c6080-07b8-11ea-9c10-3c3e66988542.gif)
A quick and dirty fix would be to increase texture resolution to 1:1, but I think the issue is deeper in the heatmap implementation. This is heatmap with 1:1 resolution: 
![heatmap_aliasing_fix](https://user-images.githubusercontent.com/55925868/68945441-1459e000-07b9-11ea-99b4-0eddb94fb570.gif)
Related pull request: #8975
